### ISSUE TITLE
[C++20] [Modules] Diagnose exposure for TUlocal entity 

### ISSUE THREAD

**Issue by ChuanqiXu9** [llvm/llvm-project#112294]
The wording is at http://eel.is/c++draft/basic.link#14

The simplest example is:

```
export module a;
static int a = 43;
export inline int func() {
    return a;
}
```

According to the spec, we should reject the example and say “the TU-local entity ‘a’ is an exposure”.  But we didn’t implement this. We would accept the example silently. The practical affect this is, user may get weird diagnostic informations triggered by other places. Or, in the worst case, they may get runtime errors if the static variable have dyanmic initializers. 

The reason why I didn’t implement this is, to implement this now, we may need to implement something more complex than the existing mechanism for `used` information in Sema. I feel it is too complex. So I feel maybe we can implement it in Serializer, not in Sema. The idea is, if we write anythiing internal to the current TU to the BMI, we can issue a diagnostic. But it requires we make the reduced BMI (https://clang.llvm.org/docs/StandardCPlusPlusModules.html#reduced-bmi) the default. Since the current full BMI will contain the full information after all. And according to my plan, the reduced BMI will be the default in clang21. So this feature is expected to be implemented that time.

**Comment by h-vetinari** [comment]
> And according to my plan, the reduced BMI will be the default in clang21. So this feature is expected to be implemented that time.

@ChuanqiXu9, we're a few weeks away from branching for clang 21. Are you still planning to tackle this issue for that release? It's the only remaining issue in #112295, which came up in discussion about setting the FTM: #71364

**Comment by ChuanqiXu9** [comment]
> > And according to my plan, the reduced BMI will be the default in clang21. So this feature is expected to be implemented that time.
> 
> [@ChuanqiXu9](https://github.com/ChuanqiXu9), we're a few weeks away from branching for clang 21. Are you still planning to tackle this issue for that release? It's the only remaining issue in [#112295](https://github.com/llvm/llvm-project/issues/112295), which came up in discussion about setting the FTM: [#71364](https://github.com/llvm/llvm-project/issues/71364)

No. I think we would implement this feature with reduced BMI, which will be enabled by default in 22.x by our plan. So it may not be in 21.

**Comment by Bigcheese** [comment]
I'm not a fan of implementing this as part of serialization. It would mean not getting any diagnostics if you're not building a reduced PCM or just not building a PCM at all (which IDEs will do when editing that file). In general, the only errors that should happen during writing an AST file are file system errors or exceeding implementation limits. If Clang can successfully build an AST it should be able to serialize it.

I think we should track exposures in Sema and handle it there. This will also make it easier to handle the `static inline` case that hasn't been resolved in the standard yet (see https://wg21.link/p2808 and linked papers).

I believe @iains had an approach that handled this during Sema?

**Comment by ChuanqiXu9** [comment]
Handling this in Sema has a big concerning for performance. To make this we have to do a reachable analysis in AST, it is hard to implement and possibly slow.

Reduced BMI will be the default. I think we don't need to care the case we don't need to build the BMI. Since the intention of the design is to not break the isolationism of internal linkage entities. And if we don't build the BMI, it won't break the property for internal linkage entities.

The thought on layering makes sense. Yes it is better to make the serializer to do serialization only. A possible idea is polymorphism. We can have a pure serializer. Then we can inherit the serializer and perform the analysis. It will be efficient.

And I still want to do it in serializer, it is much simpler to implement and may be the most efficient method as far as I know. We can refactor them later. 

Please don't be nervous. Code can be changed. Especially the patch to implement it in serializer will be small. The important thing is, now it is 2025 and the feature is declared in 2020. People have been disappointing. It is urgent.

**Comment by ChuanqiXu9** [comment]
I changed my mind after I work on it deeply. It seems like it is not so hard to implement it in Sema. But I need to try it. If we can implement it in Sema, I'll try to implement it in Sema first. If there is performance issue, it won't be too hard to move the implementation to Serializer then.

**Comment by ChuanqiXu9** [comment]
> [@ChuanqiXu9](https://github.com/ChuanqiXu9) Did you accidentally push the change to main?

I pushed it intentionally as the maintainer of modules. I generally did so if I believe this should be done (as it is part of standard) and I have belief in the implementation. Generally I would send RFC if I am not sure if we should do it if it has a wider impact (e.g., https://discourse.llvm.org/t/rfc-c-modules-stop-using-abbrev-and-drop-the-maintainance/87063/22) or I don't have strong confidence for it (e.g, https://github.com/llvm/llvm-project/pull/92083).  And generally if something bad happens, we can revert it. Or if something minor happens, we can fix them as you pointed in the commit. This is what I do for the past years. I think it works well and efficiently.

For the patch itself, I feel it is good or at least not bad since almost the changes live in SemaModule.cpp and the changes in other parts are trivial.

**Comment by h-vetinari** [comment]
@ChuanqiXu9, what's the status here? Now that the reduced BMI is the default, is there anything still blocking the resolution of this?

**Comment by ChuanqiXu9** [comment]
> [@ChuanqiXu9](https://github.com/ChuanqiXu9), what's the status here? Now that the reduced BMI is the default, is there anything still blocking the resolution of this?

There is nothing changed since the last comment in https://github.com/llvm/llvm-project/issues/54048 

I didn't continue since I feel I don't have a plain and trivial implementation model for the remaining part... 

BTW, currently the support of modules in clang should be workable in linux except few bugs (users said there bugs but they are not reporting)
